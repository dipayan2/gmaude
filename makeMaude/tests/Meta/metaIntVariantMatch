#!/bin/sh

MAUDE_LIB=$srcdir/../../src/Main
export MAUDE_LIB

../../src/Main/maude \
  < $srcdir/metaIntVariantMatch.maude -no-banner -no-advise \
  > metaIntVariantMatch.out 2>&1

diff $srcdir/metaIntVariantMatch.expected metaIntVariantMatch.out > /dev/null 2>&1
