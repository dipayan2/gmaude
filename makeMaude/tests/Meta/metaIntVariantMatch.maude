set show timing off .

***
*** Test variant unification in the meta-interpreter.
***

load metaInterpreter

fmod XOR is
  sort XOR .

  sort Elem .
  ops c1 c2 c3 c4 : -> Elem .

  subsort Elem < XOR .
  op _+_ : XOR XOR -> XOR [ctor assoc comm] .
  op 0 : -> XOR .
  op a : -> XOR .

  vars X Y : XOR .

  eq Y + 0 = Y [variant] .
  eq X + X = 0 [variant] .
  eq X + X + Y = Y [variant] .
endfm

mod VARIANT-MATCH-TEST is
  pr META-INTERPRETER .
  sort MsgList .
  subsort Msg < MsgList .
  op me : -> Oid .
  op User : -> Cid .
  op problem:(_,_) : MatchingProblem TermList -> Attribute .
  op soln:_ : Nat -> Attribute .
  op result : MsgList -> Attribute .
  op _,_ : MsgList MsgList -> MsgList [assoc id: nil] .
  op nil : -> MsgList .

  vars X Y Z : Oid .
  var AS : AttributeSet .
  var N : Nat .
  var V : Term .
  var TL : TermList .
  var S : Substitution .
  var Q : Qid .
  var B : Bool .
  var P : Parent .
  var R : RewriteCount .
  var ML : MsgList .
  var MP : MatchingProblem .

  rl < X : User | AS > createdInterpreter(X, Y, Z) =>
     < X : User | AS > insertModule(Z, X, upModule('XOR, true)) .

  rl < X : User | AS, problem:(MP, TL) > insertedModule(X, Y) =>
     < X : User | AS, problem:(MP, TL), soln: 1, result(nil) > getVariantMatcher(Y, X, 'XOR, MP, TL, '#, none, 0) .

  rl < X : User | AS, problem:(MP, TL), soln: N, result(ML) > gotVariantMatcher(X, Y, R, S) =>
     < X : User | AS,  problem:(MP, TL), soln: (N + 1), result(ML, gotVariantMatcher(X, Y, R, S)) >
     getVariantMatcher(Y, X, 'XOR, MP, TL, '#, none, N) .
endm

erew in VARIANT-MATCH-TEST : <> 
< me : User | problem: ('_+_['X:XOR, 'c1.Elem] <=? '_+_['Y:XOR, 'c2.Elem], empty) > 
createInterpreter(interpreterManager, me, none) .

variant match in XOR : X:XOR + c1 <=? Y:XOR + c2 .

erew in VARIANT-MATCH-TEST : <> 
< me : User | problem: ('_+_['X:XOR, 'Y:XOR] <=? '_+_['c1.Elem, 'c2.Elem], empty) > 
createInterpreter(interpreterManager, me, none) .

variant match in XOR : X:XOR + Y:XOR  <=? c1 + c2 .
