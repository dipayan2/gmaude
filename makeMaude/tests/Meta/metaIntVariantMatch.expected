==========================================
erewrite in VARIANT-MATCH-TEST : <> < me : User | problem:('_+_['X:XOR,
    'c1.Elem] <=? '_+_['Y:XOR, 'c2.Elem], empty) > createInterpreter(
    interpreterManager, me, none) .
rewrites: 8
result Configuration: <> < me : User | soln: 2, result(gotVariantMatcher(me,
    interpreter(0), 3, 
  'X:XOR <- '_+_['c1.Elem, 'c2.Elem, 'Y:XOR])), problem:('_+_['X:XOR, 'c1.Elem]
    <=? '_+_['Y:XOR, 'c2.Elem], empty) > noSuchResult(me, interpreter(0), 0,
    true)
==========================================
variant match in XOR : X + c1 <=? Y + c2 .
rewrites: 3

Matcher 1
X --> c1 + c2 + Y

No more matchers.
==========================================
erewrite in VARIANT-MATCH-TEST : <> < me : User | problem:('_+_['X:XOR, 'Y:XOR]
    <=? '_+_['c1.Elem, 'c2.Elem], empty) > createInterpreter(
    interpreterManager, me, none) .
rewrites: 29
result Configuration: <> < me : User | soln: 9, result(gotVariantMatcher(me,
    interpreter(0), 10, 
  'X:XOR <- 'c1.Elem ; 
  'Y:XOR <- 'c2.Elem), gotVariantMatcher(me, interpreter(0), 0, 
  'X:XOR <- 'c2.Elem ; 
  'Y:XOR <- 'c1.Elem), gotVariantMatcher(me, interpreter(0), 0, 
  'X:XOR <- '0.XOR ; 
  'Y:XOR <- '_+_['c1.Elem, 'c2.Elem]), gotVariantMatcher(me, interpreter(0), 0,
    
  'X:XOR <- '_+_['c1.Elem, 'c2.Elem] ; 
  'Y:XOR <- '0.XOR), gotVariantMatcher(me, interpreter(0), 0, 
  'X:XOR <- '_+_['c1.Elem, '#1:XOR] ; 
  'Y:XOR <- '_+_['c2.Elem, '#1:XOR]), gotVariantMatcher(me, interpreter(0), 0, 
  'X:XOR <- '_+_['c2.Elem, '#1:XOR] ; 
  'Y:XOR <- '_+_['c1.Elem, '#1:XOR]), gotVariantMatcher(me, interpreter(0), 0, 
  'X:XOR <- '_+_['c1.Elem, 'c2.Elem, '#1:XOR] ; 
  'Y:XOR <- '#1:XOR), gotVariantMatcher(me, interpreter(0), 0, 
  'X:XOR <- '#1:XOR ; 
  'Y:XOR <- '_+_['c1.Elem, 'c2.Elem, '#1:XOR])), problem:('_+_['X:XOR, 'Y:XOR]
    <=? '_+_['c1.Elem, 'c2.Elem], empty) > noSuchResult(me, interpreter(0), 0,
    true)
==========================================
variant match in XOR : X + Y <=? c1 + c2 .
rewrites: 10

Matcher 1
X --> c1
Y --> c2

Matcher 2
X --> c2
Y --> c1

Matcher 3
X --> 0
Y --> c1 + c2

Matcher 4
X --> c1 + c2
Y --> 0

Matcher 5
X --> c1 + #1:XOR
Y --> c2 + #1:XOR

Matcher 6
X --> c2 + #1:XOR
Y --> c1 + #1:XOR

Matcher 7
X --> c1 + c2 + #1:XOR
Y --> #1:XOR

Matcher 8
X --> #1:XOR
Y --> c1 + c2 + #1:XOR

No more matchers.
Bye.
